File: MinMeetingRooms.java
//Time Complexity: O(n logn + n) = O(n logn) for sorting and going over interval length
//Space Compexity: O(n), where n is the interval length (for priority queue)
//Did it run successfully on leetcode: Yes
//Did you face any problem coding this challenge: No

class MinMeetingRooms {
    public int minMeetingRooms(int[][] intervals) {

        //sort the start time of the intervals
        Arrays.sort(intervals, (a, b) -> (a[0] - b[0]));

        //Priority queue to maintain the least element at the top of the stack always.

        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();

        for(int[] interval: intervals) {

            //Check if the priority queue is not empty and the top element in the stack is less than the current interval's start time and remove that element if true (if there is no overlapping between two intervals, we can merge those intervals and assign the same meeting room.
            if(!pq.isEmpty() && pq.peek() <= interval[0]) {
                pq.poll();
            }
            //add the current element's end interval time to the priority queue
            pq.add(interval[1]);
        }

        //return the size of the priority queue
        return pq.size();
    }
}

File: minMeetingRooms.py
# Time Complexity: O(nlog(n))
# Space Complexity: O(n)
class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """

        start= sorted([i[0] for i in intervals])
        print(start)
        end = sorted([i[1] for i in intervals])
        print(end)
        s=0
        e=0
        count =0
        ans=0

        while (s<len(intervals)):
            if start[s] <end[e]:
                count +=1
                s+=1
            else:
                count-=1
                e+=1
            ans = max(ans,count)
        return(ans)
